File: examples/Objective-C/taskvent.m
/* Task ventilator - sends task batch to workers via PUSH socket. */
#import <Foundation/Foundation.h>
#import "ZMQObjC.h"
#import "ZMQHelper.h"

int
main(void)
{
	NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
	ZMQContext *ctx = [[[ZMQContext alloc] initWithIOThreads:1U] autorelease];
	ZMQSocket *sender = [ctx socketWithType:ZMQ_PUSH];
	[sender bindToEndpoint:@"tcp://*:5557"];

	NSLog(@"Press Enter when the workers are ready: ");
	(void)getchar();
	NSLog(@"Sending tasks to workers...");

	/* Signal batch start with message of "0". */
	NSData *signalData = [@"0" dataUsingEncoding:NSUTF8StringEncoding];
	[sender sendData:signalData withFlags:0];

	/* Initialize random number generator. */
	(void)srandom((unsigned)time(NULL));

	/* Send kTaskCount tasks. */
	unsigned long totalMsec = 0UL;
	static const int kTaskCount = 100;
	for (int task = 0; task < kTaskCount; ++task) {
		/* Random workload from 1 to 100 msec. */
		int workload = within(100) + 1;
		totalMsec += workload;
		NSString *text = [NSString stringWithFormat:@"%d", workload];
		NSData *textData = [text dataUsingEncoding:NSUTF8StringEncoding];
		[sender sendData:textData withFlags:0];
	}
	NSLog(@"Total expected cost: %lu ms", totalMsec);

	/* Let IOThreads finish sending. */
	sleep(1);

	[sender close];
	[pool drain];
	return EXIT_SUCCESS;
}

File: examples/GoCZMQ/taskvent.go
//
//   Task ventilator
//   Binds PUSH socket to tcp://localhost:5557
//   Sends batch of tasks to workers via that socket
//   Initial Commit in examples/Go bay Aaron Raddon
//   Ported to goczmq by Michael Guldan github.com/michaelcoyote
//   Requires: http://github.com/zeromq/goczmq
//
package main

import (
	"fmt"
	goczmq "github.com/zeromq/goczmq"
	"math/rand"
	"time"
)

func main() {

	// Socket to send messages On
	sender := goczmq.NewSock(goczmq.Push)
	defer sender.Destroy()
	sender.Bind("tcp://*:5557")

	//  Socket to send start of batch message on
	sink := goczmq.NewSock(goczmq.Push)
	defer sink.Destroy()
	sink.Connect("tcp://localhost:5558")

	fmt.Print("Press Enter when the workers are ready: ")

	var line string
	fmt.Scanln(&line)

	fmt.Println("Sending tasks to workers…")

	sink.SendMessage([][]byte{[]byte("0")})

	// Seed the random number generator
	rand.Seed(time.Now().UnixNano())

	total_msec := 0

	for i := 0; i < 100; i++ {
		workload := rand.Intn(100)
		total_msec += workload
		msg := []byte(fmt.Sprintf("%d", workload))
		sender.SendMessage([][]byte{msg})
	}

	fmt.Printf("Total expected cost: %d msec\n", total_msec)

	time.Sleep(1e9) //  Give 0MQ time to deliver: one second ==  1e9 ns

}
